### PR TITLE
TypeScript: Improved types for UA, Session and Constants

### PR DESCRIPTION
**Pull request by fgladisch**
Hi,

This PR improves the types for SIP.js in the following ways:

- Uses real EventEmitter class from node.js "events" module to extend classes, to reflect the real implementation.
- Adds [all causes](https://sipjs.com/api/0.12.0/causes/#session-causes) to the constants ("C") namespace.
- Adds missing return types to various functions, which has lead to compiler errors in projects with strict tsconfig.json configurations.
- Adds peerConnection to SessionDescriptionHandlers as described in [in this SIP.js guide](https://sipjs.com/guides/attach-media/).
- Inital types for [ServerContext](https://sipjs.com/api/0.12.0/context/server/)
- Adds the field "id" to session.
- Fixes the signature of session.dtmf() to the [documented behaviour](https://sipjs.com/api/0.12.0/session/#dtmftone-options).

We use SIP.js at [sipgate](https://github.com/sipgate) and would like to contribute more types in the future, so feedback is welcome.

Thanks.

### CODE REVIEW

**Comment by egreenmachine** [comment]
This looks great! Had some small feedback items that I commented on. If those get addressed we can absolutely take this

**Comment by fgladisch** [comment]
I can't find your comments. What did you comment on?

**Comment by egreenmachine** [comment]
You should be able to see it now.

Just as a heads up, we are actively transitioning the entirety of the code to typescript/es6. This will hopefully make the types folder unnecessary, but at minimum the new code will match what is in the types folder at the time of the transition getting onto master.

File: types/session.d.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 import { URI } from "./uri";
+import { ServerContext, C } from "./SIP";
+import { EventEmitter } from "events";
 
 /**
   * The Session interface SIP.js is providing.
   */
 export class Session extends EventEmitter {
 
+  id: string;
```
[the remark this answers is not in the document]
Comment (fgladisch) [reply types/session.d.ts]: Alright, will be removed.

File: types/session-description-handler.d.ts [hunk 2]
```diff
@@ -3,6 +3,8 @@
  * The SessionDescriptionHandler interface SIP.js is expecting.
  */
 export interface SessionDescriptionHandler {
+  peerConnection?: RTCPeerConnection;
```
[the remark this answers is not in the document]
Comment (fgladisch) [reply types/session-description-handler.d.ts]: Okay, I will remove this.

File: types/session.d.ts [hunk 3]
```diff
@@ -121,6 +133,9 @@ export class InviteServerContext extends Session {
 
   reply(options?: InviteServerContext.Options): InviteServerContext;
 
+  // Default EventEmitter on-Handler needs to be preserved
+  on(event: string, callback: (event: Event) => void): this;
```
[the remark this answers is not in the document]
Comment (fgladisch) [reply types/session.d.ts]: One way to solve this is, is to change the return type of the event handlers to "this". This is already the behaviour they reflect and the typescript compile would be satisfied. Agreed?

**Comment by fgladisch** [comment]
I adjusted the types according to your feedback. Thanks.

File: types/session-description-handler.d.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@
  * The SessionDescriptionHandler interface SIP.js is expecting.
  */
 export interface SessionDescriptionHandler {
+
```
Comment (egreenmachine) [inline types/session-description-handler.d.ts]: Can you just remove this extra newline?

Comment (fgladisch) [reply types/session-description-handler.d.ts]: Done.

**Pull request merged**